src/main/java/com/google/api/codegen/transformer/csharp/CSharpGapicSnippetsTransformer.java

```diff
@@ -0,0 +1,181 @@
+      SnippetsFileView snippets = generateSnippets(context);
+      surfaceDocs.add(snippets);
+    }
+
+    return surfaceDocs;
+  }
+
+  @Override
+  public List<String> getTemplateFileNames() {
+    return Arrays.asList(SNIPPETS_TEMPLATE_FILENAME);
+  }
+
+  private Map<MethodFields, Function<SurfaceNamer, String>> generateMethodNameMap(
+      Model model, ApiConfig apiConfig) {
+    Map<MethodFields, Function<SurfaceNamer, String>> nameMap = new HashMap<>();
+    // Multiple interfaces with the same mixin will create method name collisions.
+    // This doesn't matter as the names will be the same in all interfaces.
+    for (Interface interfaze : new InterfaceView().getElementIterable(model)) {
+      for (MethodConfig methodConfig : apiConfig.getInterfaceConfig(interfaze).getMethodConfigs()) {
+        if (methodConfig.isFlattening()) {
+          final Method method = methodConfig.getMethod();
+          List<ImmutableList<Field>> fieldss = methodConfig.getFlattening().getFlatteningGroups();
+          if (fieldss.size() > 1) {
+            for (int i = 0; i < fieldss.size(); i++) {
+              final String suffix = Integer.toString(i + 1);
+              nameMap.put(
+                  new MethodFields(method, fieldss.get(i)),
+                  new Function<SurfaceNamer, String>() {
+                    @Override
+                    public String apply(SurfaceNamer namer) {
+                      return namer.getApiMethodName(method) + suffix;
+                    }
+                  });
+            }
+          }
+        }
+      }
+    }
+    return nameMap;
+  }
+
+  private ModelTypeTable createTypeTable(String implicitPackageName) {
+    return new ModelTypeTable(
+        new CSharpTypeTable(implicitPackageName),
+        new CSharpModelTypeNameConverter(implicitPackageName));
+  }
+
+  private SnippetsFileView generateSnippets(SurfaceTransformerContext context) {
+    SurfaceNamer namer = context.getNamer();
+    String name = namer.getApiSnippetsClassName(context.getInterface());
+    SnippetsFileView.Builder snippetsBuilder = SnippetsFileView.newBuilder();
+
+    snippetsBuilder.templateFileName(SNIPPETS_TEMPLATE_FILENAME);
+    String outputPath = pathMapper.getOutputPath(context.getInterface(), context.getApiConfig());
+    snippetsBuilder.outputPath(outputPath + File.separator + name + ".g.cs");
+    snippetsBuilder.packageName(context.getApiConfig().getPackageName() + ".Snippets");
+    snippetsBuilder.name(name);
+    snippetsBuilder.clientTypeName(namer.getApiWrapperClassName(context.getInterface()));
+    snippetsBuilder.snippetMethods(generateMethods(context));
+
+    // must be done as the last step to catch all imports
+    snippetsBuilder.imports(
+        importTypeTransformer.generateImports(context.getTypeTable().getImports()));
+
+    return snippetsBuilder.build();
+  }
+
+  private List<StaticLangApiMethodView> generateMethods(SurfaceTransformerContext context) {
```

Comment:
*Everyone stand back, I have an idea!*

There are a couple issues in this PR that I would like to address with one solution. The first is that the whole pathway of `generateMethodNameMap` -> `Map<MethodFields, Function<SurfaceNamer, String>>` -> `CSharpSurfaceNamer` is somewhat hard to follow. The second is that in the snippet, the "snippets" object is passed everywhere. 

My suggestion is to create a StaticLangApiMethodSnippetView class which would hold a StaticLangApiMethodView plus whatever extra information is necessary to render each snippet beyond what is already in StaticLangApiMethodView. It would include:

* snippetMethodName
* apiClassName (instead of clientTypeName, to be consistent)
* apiVariableName
* callerResponseTypeName

Now to form snippetMethodName, just create a SymbolTable object within this method (`generateMethods`) and use it to generate unique method names based off of the result of getApiMethodName / getAsyncApiMethodName.